Fixing wrong duration report when start_time is bigger than 0

mohamad-z opened the pull request:
VideoDetails: constructor

When start_time is not zero, 
the duration returned by ffprobe's video stream is not relative to start time!
With this modification, 
everything is alright now.

SteveJobzniak commented:
Thanks a lot for finding this, but the code is not safe to merge. It needs error checking and rewrites.

But right now I am more concerned about the start_time field. What kinda video uses that? I doubt that Instagram lets such videos be uploaded (they would never be sent by the real app). So we may need some code that says "a video has a start_time > 0, reject it and mark it as invalid".

Furthermore, I wonder what happens if you use a start_time based video as input for ffmpeg? For example, if a video is 12 seconds and start_time is at 10 seconds, does ffmpeg produce a 2 second final output?

All of this is complex....... we can't merge this. It need smore research and I don't think anyon ehere has time for that. I for sure do not have time or energy... hmm.

mohamad-z commented:
Hey Steve,
Actually I must thank you for this precious work.

Unfortunately whatever I tried to make composer codestyle command to work, It did not and gave me 
sh: 1: lazydoctor: not found
Script lazydoctor -c composer.json -pfo >/dev/null handling the codestyle event returned with error code 127

error.
I have no idea what lazydoctor is and could not find anything about it on the internet!
i even executed
php-cs-fixer fix --config=.php_cs.dist --allow-risky yes
manually but then executing
php devtools/checkStyle.php x

gave me 

PHP Warning:  require(/mnt/backup/MarsWorkplace/Instagram-API/devtools/../vendor/autoload.php): failed to open stream: No such file or directory in /mnt/backup/MarsWorkplace/Instagram-API/devtools/checkStyle.php on line 6
PHP Fatal error:  require(): Failed opening required '/mnt/backup/MarsWorkplace/Instagram-API/devtools/../vendor/autoload.php' (include_path='.:/usr/share/php') in /mnt/backup/MarsWorkplace/Instagram-API/devtools/checkStyle.php on line 6

error and warnings!

So I gave up making things any prettier!
Any advise on that?

mohamad-z commented:
Please also take a look at this log message:

```
"streams": [
        {
            "index": 0,
            "codec_name": "h264",
            "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
            "profile": "High",
            "codec_type": "video",
            "codec_time_base": "1/50",
            "codec_tag_string": "avc1",
            "codec_tag": "0x31637661",
            "width": 640,
            "height": 360,
            "coded_width": 640,
            "coded_height": 360,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p",
            "level": 30,
            "chroma_location": "left",
            "refs": 1,
            "is_avc": "true",
            "nal_length_size": "4",
            "r_frame_rate": "25/1",
            "avg_frame_rate": "25/1",
            "time_base": "1/12800",
            "start_pts": 655872,
            "start_time": "51.240000",
            "duration_ts": 1293312,
            "duration": "101.040000",
            "bit_rate": "1215662",
            "bits_per_raw_sample": "8",
            "nb_frames": "1245",
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "VideoHandler"
            }
        },
        {
            "index": 1,
            "codec_name": "aac",
            "codec_long_name": "AAC (Advanced Audio Coding)",
            "profile": "LC",
            "codec_type": "audio",
            "codec_time_base": "1/44100",
            "codec_tag_string": "mp4a",
            "codec_tag": "0x6134706d",
            "sample_fmt": "fltp",
            "sample_rate": "44100",
            "channels": 2,
            "channel_layout": "stereo",
            "bits_per_sample": 0,
            "r_frame_rate": "0/0",
            "avg_frame_rate": "0/0",
            "time_base": "1/44100",
            "start_pts": 2256288,
            "start_time": "51.162993",
            "duration_ts": 4452777,
            "duration": "100.970000",
            "bit_rate": "133370",
            "nb_frames": "2145",
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0
            },
            "tags": {
                "language": "eng",
                "handler_name": "SoundHandler"
            }
        }
    ],
    "format": {
        "filename": "/mnt/backup/MarsWorkplace/telegram-youtube-site/App/Control/../Content/video/wm.401a10c2f659f77db18d07b313932db4-640x360-01.mp4",
        "nb_streams": 2,
        "nb_programs": 0,
        "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
        "format_long_name": "QuickTime / MOV",
        "start_time": "51.162993",
        "duration": "49.807000",
        "size": "8439883",
        "bit_rate": "1355613",
        "probe_score": 100,
        "tags": {
            "major_brand": "isom",
            "minor_version": "512",
            "compatible_brands": "isomiso2avc1mp41",
            "encoder": "Lavf57.50.100"
        }
    }
```
When I was trying to post a video file that I believed is shorter than 60 seconds,  get long video duration error! I produced above log output and realized what's the problem.
After I made the committed changed, video successfully uploaded and played nice in Instagram.

So I don't think it is a good idea to ignore these kind of video files.
Also I think start_time > 0 might be the result of some cropping or other manipulations on the video. I'm not sure though!

SteveJobzniak commented:
@mohamad-z About the lazydoctor question, it's part of https://github.com/SteveJobzniak/LazyJsonMapper

So I guess you're not patching the master-branch of Instagram-API.

The way to do that would be to install `composer require mgp25/instagram-php dev-master` to get the latest code. Then go into the `vendor/mgp/instagram-php` folder and type `composer update -o` to install all libraries.

After that, `composer codestyle` will work inside the `vendor/mgp/instagram-php` folder and development can begin.

SteveJobzniak commented:
@mohamad-z Thanks for the log. That's a massive start_time! So that video file has 51 seconds of dead data at the start that never gets played.

From what I can see, start_time is basically telling the video player to ignore earlier data in the video and pretend that the video is shorter.

Anyway, what I said above regarding this is that we must probably detect `if (start_time > 0)` and block those videos from even being uploaded, because I do not think Instagram.com expects them (I doubt that Instagram's real app creates video files with start_time metadata). And we block all uploads of impossible data that a real app would never create.

Then, we also must research how ffmpeg re-encodes/changes container on start_time videos (hopefully it deletes the earlier seconds of the videos and obeys the start_time command).

And lastly, this patch needs lots of rewriting to make it safe. It assumes the start_time field exists, but it surely won't exist in some container formats. And the big NOTE block can be summed up in a much shorter way. And the `floatval()` is a needless function call and is better rewritten as `(float) (a+b)`.

valga commented:
@mohamad-z The main question is: what the app does with such videos? Does it see them with long or short duration? Could you share such video with us, so we can check it?

mohamad-z commented:
@valga I'm pretty sure that the app plays them pretty normal with the short duration. Unfortunately that video was for about 2 or 3 weeks ago and there is no way I could find it again. Once I fixed the code and could upload it, I removed the video and don't remember what it was.
But there is a good chance that I find something like that again. I'll update you as soon as I find one

RuslanHimikEvchev commented:
madz0, with ffmpeg you can use -avoid_negative_ts make_zero on crop video. Its a good idea, to using this option in all ffmpeg transcoding or repacks operations
